src/servers/src/grpc/flight/stream.rs

```diff
@@ -129,6 +157,54 @@ impl Stream for FlightRecordBatchStream {
     }
 }
 
+/// Column of json type in record batches needs to be serialized before sending
+/// to the client. This function changes the json column in schema to string column.
+fn change_json_column_to_string_in_schema(schema: Arc<Schema>) -> Schema {
+    let column_schemas = schema
+        .column_schemas()
+        .iter()
+        .map(|column| {
+            if column.data_type.is_json() {
+                ColumnSchema::new(
+                    column.name.clone(),
+                    ConcreteDataType::string_datatype(),
+                    column.is_nullable(),
+                )
+            } else {
+                column.clone()
+            }
+        })
+        .collect();
+    Schema::new(column_schemas)
+}
+
+/// Column of json type in record batches needs to be serialized before sending
+/// to the client. This function serializes the binary data of json type to string.
+fn serialize_json_column(schema: Arc<Schema>, recordbatch: &RecordBatch) -> RecordBatch {
+    let mut column_schemas = Vec::with_capacity(schema.column_schemas().len());
+    let mut vectors: Vec<Arc<dyn Vector>> = Vec::with_capacity(schema.column_schemas().len());
+    for (i, column) in schema.column_schemas().iter().enumerate() {
+        let vector = recordbatch.column(i);
+        if column.data_type.is_json() {
+            // Safety: json column with binary vector
+            let vector = vector.as_any().downcast_ref::<BinaryVector>().unwrap();
+            let vector = vector.from_json_vector_to_string_vector();
+            vectors.push(Arc::new(vector) as VectorRef);
+            column_schemas.push(ColumnSchema::new(
+                column.name.clone(),
+                ConcreteDataType::string_datatype(),
+                column.is_nullable(),
+            ));
+        } else {
+            vectors.push(vector.clone());
+            column_schemas.push(column.clone());
+        }
+    }
+
+    // Safety: only changed a column type from binary to string
+    RecordBatch::new(Arc::new(Schema::new(column_schemas)), vectors).unwrap()
+}
```

Comment:
How about moving this part into a UDF, we only use the grpc protocol for internal usage(e.g., frontend <-> datanode, sqlness tests). we won't use grpc protocol for queries.